### PR TITLE
[Heartbeat] States ES Client backoff and run once exit

### PR DESCRIPTION
**Pull request by emilioalvap**
## Summary

Add ES states client backoff and exit when running as `run_once` mode.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Set `run_once` mode flag to `true` and point `elasticsearch` output to a non-existing service.
2. Run heartbeat and wait for it to exit with:
```
Exiting: run_once mode fatal error: could not establish connection on 3 attempts. Last error: couldn't connect to any of the configured Elasticsearch hosts. Errors: [error connecting to Elasticsearch at http://localhost:9200: Get "http://localhost:9200": dial tcp 127.0.0.1:9200: connect: connection refused]
```

### CODE REVIEW

File: heartbeat/beater/heartbeat.go [hunk 1]
```diff
@@ -66,11 +68,20 @@ func New(b *beat.Beat, rawConfig *conf.C) (beat.Beater, error) {
...
 		// Connect to ES and setup the State loader if the output is not managed by agent
-		if err := makeStatesClient(b.Config.Output.Config(), replaceStateLoader, parsedConfig.RunFrom); err != nil {
-			logp.L().Warnf("could not connect to ES for state management during initial load: %s", err)
+		// Note this, intentionally, blocks until connected or max attempts reached
+		esClient, err := makeESClient(b.Config.Output.Config(), 3, 2*time.Second)
+		if err != nil {
+			if parsedConfig.RunOnce {
+				return nil, fmt.Errorf("run_once mode fatal error: %w", err)
+			} else {
+				logp.L().Warn(err)
```
Comment (vigneshshanmugam) [inline heartbeat/beater/heartbeat.go:81]: this causes the error to be chatty as its already logged by libbbeat, better to ignore it.

File: heartbeat/beater/heartbeat.go [hunk 2]
```diff
@@ -253,8 +264,12 @@ func (bt *Heartbeat) RunCentralMgmtMonitors(b *beat.Beat) {
 			return nil
 		}
 
-		if err := makeStatesClient(outCfg.Config(), bt.replaceStateLoader, bt.config.RunFrom); err != nil {
+		// Backoff panics with 0 duration, set to smallest unit
+		esClient, err := makeESClient(outCfg.Config(), 1, 1*time.Nanosecond)
+		if err != nil {
 			logp.L().Warnf("could not connect to ES for state management during managed reload: %s", err)
+		} else {
+			bt.replaceStateLoader(monitorstate.MakeESLoader(esClient, "synthetics-*,heartbeat-*", bt.config.RunFrom))
```
Comment (vigneshshanmugam) [inline heartbeat/beater/heartbeat.go:272]: nit: we can move those index patterns inside state package as constants.

File: heartbeat/beater/heartbeat.go [hunk 3]
```diff
@@ -298,16 +313,27 @@ func (bt *Heartbeat) Stop() {
...
+		wait,
+		wait,
+	)
 
-	if esClient != nil {
-		logp.L().Info("using ES states loader")
-		replace(monitorstate.MakeESLoader(esClient, "synthetics-*,heartbeat-*", runFrom))
+	for i := 0; i < attempts; i++ {
+		esClient, err = eslegclient.NewConnectedClient(cfg, "Heartbeat")
+		if err == nil {
```
Comment (vigneshshanmugam) [inline heartbeat/beater/heartbeat.go:331]: nit: connectDelay.Reset()

**Review by vigneshshanmugam — APPROVED** [review]
LGTM with minor comments.

File: heartbeat/beater/heartbeat.go (hunk 1 above)
Comment (emilioalvap) [reply heartbeat/beater/heartbeat.go:81]: ++. IMO, there should still be a warning specifying that states client could not connect and therefore will not be used, so I refactored the error message to reduce chattiness,  wdyt?

**Review by vigneshshanmugam — APPROVED** [review]
LGTM

**Pull request merged**